[CAS] Add LLVMCAS library with InMemoryCAS implementation (#114096)
llvm/llvm-project: llvm/docs/ContentAddressableStorage.md

# Content Addressable Storage

## Introduction to CAS

Content Addressable Storage, or `CAS`, is a storage system where it assigns
unique addresses to the data stored. It is very useful for data deduplicaton
and creating unique identifiers.

> Unlikely other kind of storage system like file system, CAS is immutable. It

Reviewer: Nit: 
Change Unlikely to Unlike

> Unlikely other kind of storage system like file system, CAS is immutable. It

Reviewer: ```suggestion
Unlike other kind of storage system like file system, CAS is immutable. It
```

```
struct CASObject {
  ArrayRef<char> Data;
  ArrayRef<CASObject*> Refs;
}
```

Such abstraction can allow simple composition of CASObjects into a DAG to
represent complicated data structure while still allowing data deduplication.
Note you can compare two DAGs by just comparing the CASObject hash of two
root nodes.

## LLVM CAS Library User Guide

The CAS-like storage provided in LLVM is `llvm::cas::ObjectStore`.
To reference a CASObject, there are few different abstractions provided
with different trade-offs:

### ObjectRef

`ObjectRef` is a lightweight reference to a CASObject stored in the CAS.
This is the most commonly used abstraction and it is cheap to copy/pass
along. It has following properties:

* `ObjectRef` is only meaningful within the `ObjectStore` that created the ref.
`ObjectRef` created by different `ObjectStore` cannot be cross-referenced or
compared.
* `ObjectRef` doesn't guarantee the existence of the CASObject it points to. An
explicitly load is required before accessing the data stored in CASObject.
This load can also fail, for reasons like but not limited to: object does
not exist, corrupted CAS storage, operation timeout, etc.

> * If two `ObjectRef` are equal, it is guarantee that the object they point to

Reviewer: ```suggestion
* If two `ObjectRef` are equal, it is guaranteed that the object they point to
```

```
ObjectRef A, B, C;
Expected<ObjectRef> Stored = ObjectStore.store("data", {A, B});
Expected<ObjectProxy> Loaded = ObjectStore.getProxy(C);
```

It also provides APIs to convert between `ObjectRef`, `ObjectProxy` and
`CASID`.

## CAS Library Implementation Guide

The LLVM ObjectStore APIs are designed so that it is easy to add
customized CAS implementation that are interchangeable with builtin
CAS implementations.

To add your own implementation, you just need to add a subclass to
`llvm::cas::ObjectStore` and implement all its pure virtual methods.
To be interchangeable with LLVM ObjectStore, the new CAS implementation
needs to conform to following contracts:

* Different CASObject stored in the ObjectStore needs to have a different hash
and result in a different `ObjectRef`. Vice versa, same CASObject should have
same hash and same `ObjectRef`. Note two different CASObjects with identical
data but different references are considered different objects.
* `ObjectRef`s are comparable within the same `ObjectStore` instance, and can
be used to determine the equality of the underlying CASObjects.
* The loaded objects from the ObjectStore need to have the lifetime to be at

> least as long as the ObjectStore itself.

Author: To the last point, currently the loaded object `ObjectProxy` is not ref counted and there doesn't exist an `unload` operation for the CAS, so it is hard to define what is the last reference of the ObjectProxy. In general, this is legal currently:
```
auto Loaded = CAS.getProxy(Ref);
if (Loaded)
  return Loaded->getData(); // return the underlying buffer without extending the lifetime of the Proxy
```
Define a shorter lifetime makes sense (as a relaxed lifetime is also legal), but we should probably add some kind of error checking to catch the case in my example above to make CAS usage portable.

> least as long as the ObjectStore itself.

Reviewer 2: Hmm, that's a bit tricky then. For now updating the text to clarify those points will help a lot.

If not specified, the behavior can be implementation defined. For example,
`ObjectRef` can be used to point to a loaded CASObject so
`ObjectStore` never fails to load. It is also legal to use a stricter model
than required. For example, an `ObjectRef` that can be used to compare
objects between different `ObjectStore` instances is legal but user

> of the ObjectStore should not depend on this behavior.

Author: Let me know if the newer explanation works better or is it just better left it off.

> of the ObjectStore should not depend on this behavior.

Reviewer 2: It is better, but I think it still conflicts with one of your earlier statements:
```Markdown
* `ObjectRef` is only meaningful within the `ObjectStore` that created the ref.
`ObjectRef` created by different `ObjectStore` cannot be cross-referenced or
compared.
```

Maybe some of the document just needs to be reorganized to present this in a more cohesive way? Or perhaps you just need to soften the above to say "typically" or something to convey that its not a requirement?

# Content Addressable Storage

## Introduction to CAS

> Content Addressable Storage, or `CAS`, is a storage system where it assigns

Reviewer 3: ```suggestion
Content Addressable Storage, or `CAS`, is a storage system that assigns
```

unique addresses to the data stored. It is very useful for data deduplicaton
and creating unique identifiers.

> Unlike other kinds of storage system like a file system, CAS is immutable. It

Reviewer 3: ```suggestion
Unlike other kinds of storage systems like file systems, CAS is immutable. It
```

is more reliable to model a computation by representing the inputs and outputs
of the computation using objects stored in CAS.

The basic unit of the CAS library is a CASObject, where it contains:

* Data: arbitrary data
* References: references to other CASObject

It can be conceptually modeled as something like:

```
struct CASObject {
  ArrayRef<char> Data;
  ArrayRef<CASObject*> Refs;
}
```

> With this abstraction, it is possible to compose CASObjects into a DAG that is

Reviewer 3: seems inconsistent that CASObject is not quoted in backticks?
